backtesting/backtesting.py

```diff
@@ -1221,101 +1236,172 @@ def maximize(stats: pd.Series, _key=maximize):
+            finally:
+                del Backtest._mp_backtests[backtest_uuid]
 
-        if return_heatmap:
-            return self._results, heatmap
-        return self._results
+            best_params = heatmap.idxmax()
 
-    @staticmethod
+            if pd.isnull(best_params):
+                # No trade was made in any of the runs. Just make a random
+                # run so we get some, if empty, results
+                self.run(**param_combos[0])  # type: ignore
+            else:
+                # Re-run best strategy so that the next .plot() call will render it
+                self.run(**dict(zip(heatmap.index.names, best_params)))
+
+            if return_heatmap:
+                return self._results, heatmap
+            return self._results
+
+        def _optimize_skopt() -> Union[pd.Series, Tuple[pd.Series, pd.Series]]:
+            """
+            Optimize strategy parameters to an optimal combination using
+            scikit-optimize. Returns result `pd.Series` of
+            the best run.
+
+            `maximize` is a string key from the
+            `backtesting.backtesting.Backtest.run`-returned results series,
+            or a function that accepts this series object and returns a number;
+            the higher the better. By default, the method maximizes
+            Van Tharp's [System Quality Number](https://google.com/search?q=System+Quality+Number).
+
+            Additional keyword arguments represent strategy arguments with
+            list-like collections of possible values. For example, the following
+            code finds and returns the "best" of the 7 admissible (of the
+            9 possible) parameter combinations:
+
+                backtest.optimize(sma1=[5, 10, 15], sma2=[10, 20, 40],
+                                constraint=lambda p: p.sma1 < p.sma2)
+
+            .. TODO::
+                Improve multiprocessing with joblib.parallel .
+            """
+            dimensions = []
+            for key, value in kwargs.items():
+                args_array = np.array(value)
+
+                if args_array.dtype.kind == 'i':
+                    dimensions.append(Integer(name=key, low=min(value), high=max(value)))
+                elif args_array.dtype.kind == 'f':
+                    dimensions.append(Real(name=key, low=min(value), high=max(value)))
+                else:
+                    dimensions.append(Categorical(args_array.tolist(), name=key))
+
+            # @ use_named_args(dimensions=dimensions)
+            # def convert(n1, n2):
+            #     res = self.run(**dict(zip(['n1', 'n2'], [n1, n2])))
+            #     return -res[maximize_key]
+            @ use_named_args(dimensions=dimensions)
+            def convert(**params):
+                index = []
+                for item in params:
+                    index.append("'" + item + "'")
+                res = self.run(**dict(zip(index, params)))
```

Comment:
Given [what `use_named_args()` does to params](https://github.com/scikit-optimize/scikit-optimize/blob/de32b5fd2205a1e58526f3cacd0422a26d315d0f/skopt/utils.py#L800-L803), we can just do
```py
res = self.run(**params)
```
?

Reply:
And since we haven't checked params values with `constraint=` previously, we should check that they pass it now before `self.run()`.

Reply:
Yes, sorry! I am still working on this part. The integer n1 and n2 doesn't easily generalize to a case with multiple features each of different kind with the current input format for forest_minimize(). Once I fix this generalization (I need to go in more detail) I will apply your other feedbacks as well. Thanks...

Reply:
Applied checking that params pass `constraint=`. Additionally, we're waiting for https://github.com/scikit-optimize/scikit-optimize/pull/971. Welcome a review.